### PR TITLE
Robustness Allow errors and partial responses from deterministic model

### PR DESCRIPTION
**Pull request by serathius**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by serathius** [comment]
cc @ahrtr @jmhbnz @ptabor @chaochn47

File: tests/robustness/model/history.go [hunk 1]
```diff
@@ -364,38 +364,38 @@ func rangeResponse(kvs []*mvccpb.KeyValue, count int64, revision int64) EtcdNonD
...
 
-func failedResponse(err error) EtcdNonDeterministicResponse {
-	return EtcdNonDeterministicResponse{Err: err}
+func failedResponse(err error) MaybeEtcdResponse {
+	return MaybeEtcdResponse{Err: err}
 }
 
-func unknownResponse(revision int64) EtcdNonDeterministicResponse {
-	return EtcdNonDeterministicResponse{ResultUnknown: true, EtcdResponse: EtcdResponse{Revision: revision}}
+func unknownResponse(revision int64) MaybeEtcdResponse {
```
Comment (chaochn47) [inline tests/robustness/model/history.go:374]: How about renaming this function to `partialResponse()` in the follow ups?

**Comment by chaochn47** [comment]
Not sure what's the motivation of the PR, thanks for clarification of the definition of deterministic and non-deterministic model.

LGTM.

**Review by chaochn47 — APPROVED** [review]

File: tests/robustness/model/history.go (hunk 1 above)
Comment (serathius) [reply tests/robustness/model/history.go]: Done.

File: tests/robustness/model/non_deterministic.go [hunk 2]
```diff
@@ -90,18 +87,26 @@ func (states nonDeterministicState) stepFailedRequest(request EtcdRequest) nonDe
...
-// stepSuccessfulRequest filters possible states by leaving ony states that would respond correctly.
-func (states nonDeterministicState) stepSuccessfulRequest(request EtcdRequest, response EtcdNonDeterministicResponse) nonDeterministicState {
+// stepPartialRequest filters possible states by leaving ony states that would return proper revision.
+func (states nonDeterministicState) stepPartialRequest(request EtcdRequest, responseRevision int64) nonDeterministicState {
 	newStates := make(nonDeterministicState, 0, len(states))
 	for _, s := range states {
-		newState, gotResponse := s.step(request)
-		if Match(EtcdNonDeterministicResponse{EtcdResponse: gotResponse}, response) {
+		newState, modelResponse := s.step(request)
+		if modelResponse.Revision == responseRevision {
```
[the remark this answers is not in the document]
Comment (serathius) [reply tests/robustness/model/non_deterministic.go:95]: It's explained in deterministic model doc

[the remark this answers is not in the document]
Comment (serathius) [reply tests/robustness/model/non_deterministic.go:95]:
```
// Model can still respond with error or incomplete response.
//   - Error for etcd known errors, like future revision or compacted revision.
//   - Incomplete response when requests is correct, but model doesn't have all
//     to provide a full response. For example stale reads as model doesn't store
//     whole change history as real etcd does.
```
It's not yet implemented, but I added the comment to provide the intention.

File: tests/robustness/model/deterministic.go [hunk 3]
```diff
@@ -25,7 +25,18 @@ import (
...
 
-// DeterministicModel assumes that all requests succeed and have a correct response.
+// DeterministicModel assumes a deterministic execution of etcd requests. All
+// requests that client called where executed and persisted by etcd. This
+// assumption is good for simulating etcd behavior (aka writing a fake), but not
+// for validating correctness as requests might be lost or interrupted. It
+// requires perfect knowledge of what happened to request which is not possible
+// in real systems.
+//
+// Model can still respond with error or incomplete response.
```
Comment (jmhbnz) [inline tests/robustness/model/deterministic.go:35]: Nit - I would prefer to align the language around partial / incomplete. Code makes heavy use of word partial so suggest comment stick to that term.

```suggestion
// Model can still respond with error or partial response.
```

**Review by jmhbnz — APPROVED** [review]
LGTM - Two minor nitpicks that don't impact code function, happy for these to be addressed at a later date.

File: tests/robustness/model/deterministic.go (hunk 3 above)
Comment (serathius) [reply tests/robustness/model/deterministic.go:35]: Done

**Pull request merged**